lib/rspec/mocks/proxy.rb

```diff
@@ -406,34 +406,66 @@ class PartialClassDoubleProxy < PartialDoubleProxy
     # @private
     class ProxyForNil < PartialDoubleProxy
       def initialize(order_group)
-        @warn_about_expectations = true
+        set_expectation_behavior
         super(nil, order_group)
       end
 
+      attr_accessor :disallow_expectations
       attr_accessor :warn_about_expectations
       alias warn_about_expectations? warn_about_expectations
 
       def add_message_expectation(method_name, opts={}, &block)
-        warn(method_name) if warn_about_expectations?
+        warn_or_raise(method_name) if warn_or_disallow_expectations?
         super
       end
 
       def add_negative_message_expectation(location, method_name, &implementation)
-        warn(method_name) if warn_about_expectations?
+        warn_or_raise(method_name) if warn_or_disallow_expectations?
         super
       end
 
       def add_stub(method_name, opts={}, &implementation)
-        warn(method_name) if warn_about_expectations?
+        warn_or_raise(method_name) if warn_or_disallow_expectations?
         super
       end
 
     private
 
+      def set_expectation_behavior
+        case RSpec::Mocks.configuration.allow_message_expectations_on_nil
+        when false
+          @warn_about_expectations = false
+          @disallow_expectations = true
+        when true
+          @warn_about_expectations = false
+          @disallow_expectations = false
+        else
+          @warn_about_expectations = true
+          @disallow_expectations = false
+        end
+      end
+
+      def warn_or_disallow_expectations?
+        @disallow_expectations || @warn_about_expectations
+      end
+
+      def warn_or_raise(method_name)
+        if disallow_expectations
+          raise_error(method_name)
+        else
+          warn(method_name)
+        end
+      end
```

Comment:
The `@warn_about_expecations` and `@disallow_expectations` variables are checked multiple times, and I'm thinking we can clean this up considerably by introducing a polymorphic interface and 3 implementations.  I'm thinking of something like this:

``` ruby
module Notifiers
  module Null
    def self.notify(*); end
  end

  class Warn
    def initialize(error_generator)
      @error_generator = error_generator
    end

    def notify(method_name)
      msg = @error_generator.exception_on_nil_message(method_name)
      RSpec.warning(msg)
    end
  end

  class Raise
    def initialize(error_generator)
      @error_generator = error_generator
    end

    def notify(method_name)
      @error_generator.raise_expecation_on_nil_error(method_name)
    end
  end
end

def set_expectation_behavior
  @notifier = case RSpec::Mocks.configuration.allow_message_expectations_on_nil
    when false then Notifiers::Raise.new(@error_generator)
    when true  then Notiifers::Null
    else Notifiers::Warn.new(@error_generator)
  end
end
```

Then instead of calling `warn_or_raise(method_name) if warn_or_disallow_expectations?`, you can just call `@notifier.notify(method_name)` and trust it'll do the right thing.

Thoughts?